### ISSUE TITLE
Audit and fix release automation process

### ISSUE THREAD

**Issue by joshrotenberg** [joshrotenberg/mdbook-lint#191]
## Problem

The current release automation using release-plz is not working reliably. Recent issues with v0.11.4 release:

- Crates were published to crates.io but GitHub release/tag were not created
- Merging the release PR did not create the expected git tag or GitHub release
- Required manual intervention to complete the release process

## Tasks

- [ ] Investigate why release-plz is only partially completing releases
- [ ] Review current release-plz configuration in 
- [ ] Check GitHub Actions permissions and workflow triggers
- [ ] Consider if we need to adjust the release workflow or switch tools
- [ ] Document the complete release process including manual fallback steps
- [ ] Test the release process with a pre-release or dry run

## Recent Issues Encountered

1. v0.11.3 was skipped entirely (went from 0.11.2 to 0.11.4)
2. v0.11.4 was published to crates.io but not tagged/released on GitHub
3. Binary builds only trigger on GitHub release creation, not crate publication
4. Release PR merge doesn't consistently create tags/releases

## Expected Behavior

When a release PR is merged:
1. Git tag should be created and pushed
2. All crates should be published to crates.io
3. GitHub release should be created
4. Binary builds should be triggered automatically

## Current Workaround

Manual steps required:
```bash
# Create and push tag
git tag v0.11.4
git push origin v0.11.4

# Create GitHub release
gh release create v0.11.4 --title "..." --notes "..." --latest
```

## References

- PR #190 (v0.11.4 release PR that didn't fully work)
- Current  configuration
- release-plz documentation: https://release-plz.dev/

**Comment by joshrotenberg** [comment]
# Comprehensive Release-plz Configuration Analysis

After thorough investigation, I've identified several critical issues causing our release failures.

## 🔴 Critical Issues Found

### 1. **Hardcoded Version Dependencies** (PRIMARY CAUSE)
The workspace crates have hardcoded version dependencies instead of using workspace references:

```toml
# Current (WRONG) - in crates/mdbook-lint-cli/Cargo.toml
mdbook-lint-core = { version = "0.11.4", path = "../mdbook-lint-core" }
mdbook-lint-rulesets = { version = "0.11.4", path = "../mdbook-lint-rulesets" }
```

This prevents release-plz from automatically bumping internal dependency versions, causing version mismatch errors during publishing.

### 2. **Missing Workspace Dependencies Section**
The root Cargo.toml doesn't declare internal workspace crates in `[workspace.dependencies]`.

### 3. **Evidence of Failed Releases**
- Two separate commits for v0.11.4 release (77bf73e and 6f43776)
- First commit updated files, second had no changes
- Suggests the first release attempt failed partway through

## 🛠️ Immediate Fixes Required

### Fix 1: Add to root Cargo.toml
```toml
[workspace.dependencies]
# ... existing dependencies ...

# Internal workspace crates
mdbook-lint-core = { version = "0.11.4", path = "crates/mdbook-lint-core" }
mdbook-lint-rulesets = { version = "0.11.4", path = "crates/mdbook-lint-rulesets" }
```

### Fix 2: Update all crates/*/Cargo.toml files
```toml
# Change from hardcoded versions to workspace references
mdbook-lint-core = { workspace = true }
mdbook-lint-rulesets = { workspace = true }
```

### Fix 3: Update release-plz.toml
```toml
[workspace]
dependencies_update = false
publish = true
publish_timeout = "15m"  # Increased from 10m
git_release_enable = true
git_tag_enable = true
git_tag_name = "v{{ version }}"
release_always = false  # Safer for workspaces
changelog_update = false  # Explicit setting
```

### Fix 4: Update .github/workflows/release-plz.yml permissions
```yaml
permissions:
  pull-requests: write
  contents: write
  packages: write  # Add this for crates.io publishing
```

## ⚠️ Additional Issues

1. **Publishing timeout may be too short** (10m → 15m recommended)
2. **Missing `release_always = false`** which is safer for workspace projects
3. **Binary release workflow** only triggers on release creation, so if release-plz fails, binaries won't build

## ✅ What's Working Correctly

- Basic release-plz configuration structure
- Git tagging configuration
- Individual package release disabling (to avoid tag conflicts)
- GitHub Actions workflow triggers

## 📊 Root Cause Analysis

The primary failure mode is:
1. release-plz tries to update versions
2. Encounters hardcoded internal dependency versions
3. Can't resolve version conflicts during publishing
4. Partially publishes to crates.io
5. Fails to create GitHub release/tag
6. Manual intervention required

## 🔍 Historical Context

Looking at git history, there have been multiple attempts to fix release issues:
- `52b4e9b` fix: configure release-plz to avoid tag conflicts
- `eb352ce` chore: remove git-cliff configuration  
- `507abe9` fix: simplify release-plz setup to match standard Rust projects

## 📋 Action Plan

1. [ ] Implement workspace dependency fixes (Critical)
2. [ ] Update release-plz.toml configuration
3. [ ] Update GitHub Actions workflow permissions
4. [ ] Test with a dry-run or pre-release
5. [ ] Document the manual fallback process
6. [ ] Consider adding release validation steps

## 📚 References

- [release-plz workspace documentation](https://release-plz.dev/docs/workspace)
- [Cargo workspace dependencies RFC](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-dependencies-table)
- Current issues: v0.11.4 published to crates.io but required manual tag/release creation
